### PR TITLE
on="validator-update"

### PR DESCRIPTION
**Pull request by timlindvall**
The eager argument causes the validate modifier to run validation checks on initial render and when any of the arguments it depends on changes. This allows for form elements to update their validation state when the application state changes, in addition to user input.

Included two automated test cases to cover this argument and a sample in the docs. I also have a small test case I wrote in the addon's dummy app in my local Git repo - let me know if you'd like me to add it to this PR as well.

Thanks in advance for your consideration and review!

### CODE REVIEW

**Comment by timlindvall** [comment]
Hi @sukima,

Just checking back on this PR. When convenient, please let me know if you have any questions or concerns about this proposed change and the API it introduces.

Thanks! 😄

**Comment by sukima** [comment]
I have a question. If you did a `{{did-insert validate}}` instead would this feature be out of scope for this addon? In other words is the need for eager a niche enough need that a more explicit use be appropriate?

Don't get me wrong. I understand the motivation here but I have had several people tell me that solutions that use an `eager` flag tend to obfuscate what the intent is. It is a little confusing without knowing the implemementation details to understand what it does.

**Comment by timlindvall** [comment]
Hmm... well, AFAIK, relying on `did-insert`/`did-update` outside of Classic -> Octane migration cases is an anti-pattern. (Part of the effort that brought me to `ember-validity-modifer` was rethinking how to handle validation to avoid `did-insert`/`did-update`. 😅 )

This logic could be encompassed in a separate modifier. Something like...

```
<input
  {{revalidate-on-change
    @foo
    @bar
    this.baz
  }}
  {{validate this.validate}}
/>

modifier(function revalidateOnChange(element, ...arguments) {
  element.checkValidity();
});
```

But, that said, that feels really close to `did-insert`/`did-update` behavior.

I see your concern about the `eager` arg, though. It's not necessarily clear unless you've dug into modifiers and Glimmer's tracking that `eager` makes the modifier reactive to argument changes. I don't know if another name would work better? That said, I do imagine this is something that other developers might run into. Anytime you want the validation state of an input to respond to parent state, you'd need something like this.

**Comment by sukima** [comment]
Welp, I tried for a better pattern and I see now why you made it an argument to validity. Can I ask for one of the following?

1. Expose a different modifier that does the same thing as `validity` but with the *eager* semantics enabled **(preferred)**
2. Come up with a different more intention revealing name for the argument
3. Open to suggestion or if nothing better I can concede to `eager`

I did try to see if I could add a separate modifier or `{{on "validitychange" …}}` hoping I could augment the element and not the modifier but when I tried the outcome looks more ugly then I would have liked.

**Comment by timlindvall** [comment]
Sure, I'll spin up a separate branch for the first option and if that API works better I'll merge it into this PR's branch.

I also brainstormed some new argument naming ideas with teammates for the second option... some highlights...

- reactive
- dynamic
- responsive
- instant
- immediate
- onModelChange
- update
- realtime
- constant

(Thanks for the suggestions @hagmandan, @chriseppstein, @ramitha, @thousand, @thegilby!)

Also considering maybe it would work better with the current API if I instead added a special keyword to the `on` argument. Something like `on="insert-update"` or `on="install"`? (Or, those could be granular `insert` and `update` events, since the modifier is backed by a WeakSet() that can determine if it's seen a given element before. Again, maybe this gets too close to `did-insert`/`did-update`? "install" probably runs into the same problems as "eager", but it's slightly more accurate.)

**Comment by timlindvall** [comment]
Alright, I've created a separate branch that adds a new modifier `{{validity-reactive}}` that encompasses the behavior of `eager`. I've also refactored the modifier installation/uninstallation so both modifiers can share the same code. (Open to better naming suggestions for the modifier, but that's the one that jumped out to me.)

https://github.com/timlindvall/ember-validity-modifier/commit/de5dc092a2bbad7815a12bed64e25413eda676b4

This option (IMO) might make the code a bit harder to read, though. I'll defer to you whether this approach would be preferred over a renamed argument or adding special-case keyword(s) to the `on` argument. If we do want to move ahead with the separate modifier, I'll merge that branch into this PR.

Thanks!

**Comment by sukima** [comment]
You make some really good points. Do you have a preference?

**Comment by sukima** [comment]
For my own curiosity do you have examples where you needed the "eagerness"? In the use cases I've encountered data changes were always action up from the user input. This suggests that validation can be associated with external data down situations which I haven't experienced yet.

**Comment by sukima** [comment]
I kinda like the changes you made in the `validity-reactive` commit. How do you feel about it?

I also like the simplicity of the `on='insert,change,update'` but unfortunately it would cause more confusion as it mixes custom event names with DOM event names.

Lastly the term *immediate* to me feels most descriptive. Because really what we are trying to accomplish is to execute validation when the validator function changes. If I were to do this in VanillaJS™ (which this addon is based on) I would trigger a `validate` event when needed.

How about this…

```hbs
<input
  {{validity (my-custom-validator data=@dataDownThing)}}
  {{validate-on-update @dataDownThing}}
>
```

Or to do this without changes:

```hbs
<input
  {{validate (my-custom-validator @dataDownThing) on="change,custom-event"}}
  {{trigger-custom-event "custom-event" @dataDownThing}}
```

**I got it!** Simpler code no extra modifier's

```hbs
<input
  {{validity (my-custom-validator @dataDownThing) on="chnage,validator-update"}}
>
```

This way it gets the semantics of eager but is declarative on when it happens. Especially since the validator function(s) will be new references (calling the modifier's didUpdate) when their values change. This will work with your `(fn this.validate @passedArgs)` to.

**Comment by timlindvall** [comment]
Pushed a separate branch with the dummy app example I've been using to test locally (https://github.com/timlindvall/ember-validity-modifier/commit/3b280b83977730a07aed7909a040ad7a1a87ab7f). The basic use case I came up with to test against is when you want two form fields to agree with each other (say, a create password prompt).

In that example, it's possible that the validator doesn't have to handle it at all. The app code could receive the event on the first input and respond by triggering a validation event on the second. Though, I think that would be more imperative than just setting a value and having the tracking layer do the rest.

(One nuance I handwaved in the example... from a UX perspective, you might not actually want the validator to start complaining until the user tries to submit the form or until the first input event. This would likely result in a second argument - something like `@hasStartedTyping` - that flips on the first input/change/submit event and never changes back. It's possible that this is something the validity modifier could take on as its responsibility as well, if enabled by the user. But that's probably a separate discussion.)

Ooh, `on="validator-update"` LGTM. I'll post an update with a new branch... probably this evening (PDT)? Thanks!

Edit: Sorry, didn't have time to get to this today. Gonna make this update tomorrow.

**Comment by timlindvall** [comment]
Changed over the API to use `on="validator-update"`. I decided to just go ahead and add the commit on to the PR branch... lemme know if I should roll it back.

**Review by sukima — APPROVED** [review]

**Comment by sukima** [comment]
@timlindvall Do you have any reservations before this is merged?

**Comment by sukima** [comment]
> It's possible that this is something the validity modifier could take on as its responsibility as well, if enabled by the user. But that's probably a separate discussion

I thought I addressed this use case [in this example](https://github.com/sukima/ember-validity-modifier#example-rendering-validation-messages).

**Comment by timlindvall** [comment]
@sukima Nope, go ahead, thanks! 👍

**Pull request merged**
